### PR TITLE
maintenance window API

### PR DESCRIPTION
**Pull request by fspmarshall**
Core maintenance window config/api and auth-side export method.  Part of the work for the cloud agent upgrades RFD (https://github.com/gravitational/teleport/pull/20622).

### CODE REVIEW

File: api/types/maintenancewindow.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+	time.Tuesday,
+	time.Wednesday,
+	time.Thursday,
+	time.Friday,
+	time.Saturday,
+}
+
+func parseWeekday(s string) (day time.Weekday, ok bool) {
+	for _, w := range validWeekdays {
+		if strings.EqualFold(w.String(), s) || strings.EqualFold(w.String()[:3], s) {
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply api/types/maintenancewindow.go:47]: We don't _need_ to, but accepting either the full weekday name or the three-letter prefix seems to be a very common standard across scheduling APIs (e.g. systemd calendar directives, crontab).  Seems like good UX to adhere to common patterns like that.

File: lib/services/local/configuration.go [hunk 2]
```diff
@@ -444,17 +447,54 @@ func (s *ClusterConfigurationService) DeleteAllInstallers(ctx context.Context) e
...
+	authPrefix              = "authentication"
+	preferencePrefix        = "preference"
+	generalPrefix           = "general"
+	auditPrefix             = "audit"
+	networkingPrefix        = "networking"
+	sessionRecordingPrefix  = "session_recording"
+	scriptsPrefix           = "scripts"
+	uiPrefix                = "ui"
+	installerPrefix         = "installer"
+	maintenanceWindowPrefix = "maintenance-window"
```
Comment (smallinsky) [inline lib/services/local/configuration.go:499]:
```suggestion
	maintenanceWindowPrefix = "maintenance_window"
```

File: api/proto/teleport/legacy/types/types.proto [hunk 3]
```diff
@@ -5109,6 +5109,68 @@ message KubernetesResourceSpecV1 {
...
+}
+
+// AgentUpgradeWindow is the config object used to determine upcoming agent
+// upgrade windows.
+message AgentUpgradeWindow {
+  // UTCStartHour is the start hour of the maintenance window in UTC.
+  uint32 UTCStartHour = 1 [(gogoproto.jsontag) = "utc_start_hour"];
+  // Weekdays is an optional list of weekdays. If not specified, an agent upgrade window
+  // occurs every day.
+  repeated string Weekdays = 2 [(gogoproto.jsontag) = "weekdays,omitempty"];
```
Comment (smallinsky) [inline api/proto/teleport/legacy/types/types.proto:5143]: Can we use enum ?

File: api/types/maintenancewindow.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+// generator builds a closure that iterates valid maintenance window from the current day onward. Used in
+// schedule export logic and tests.
+func (w *AgentUpgradeWindow) generator(from time.Time) func() (start time.Time, end time.Time) {
+	from = from.UTC()
+	next := time.Date(
+		from.Year(),
+		from.Month(),
+		from.Day(),
+		int(w.UTCStartHour%24),
```
Comment (smallinsky) [inline api/types/maintenancewindow.go:63]: Why we need mod 24 ? how the starting hours can be larger than 23 ?

File: api/types/maintenancewindow.go [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+			for _, day := range weekdays {
+				if start.Weekday() == day {
+					return
+				}
+			}
+		}
+	}
+}
+
+// Export exports the next `n` upgarde windows as a schedule object, starting from `from`.
```
[the remark this answers is not in the document]
Comment (smallinsky) [reply api/types/maintenancewindow.go:97]: The typo seems to be still present but this commend was marked as resolved: 

```suggestion
// Export exports the next `n` upgrade windows as a schedule object, starting from `from`.
```

File: lib/auth/auth.go [hunk 6]
```diff
@@ -4192,6 +4280,80 @@ func (a *Server) SubmitUsageEvent(ctx context.Context, req *proto.SubmitUsageEve
...
+			return rsp, trace.Wrap(err)
+		}
+
+		agentWindow, ok := window.GetAgentUpgradeWindow()
+		if !ok {
+			// "unconfigured" is treated as an empty schedule value
+			return rsp, nil
+		}
+
+		sched := agentWindow.Export(time.Now(), 3)
```
Comment (smallinsky) [inline lib/auth/auth.go:4308]: Why 3 ? Can we move it to named variable ?

File: api/types/maintenancewindow_test.go [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+package types
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestMaintenceWindowAgentUpgrade(t *testing.T) {
```
Comment (smallinsky) [inline api/types/maintenancewindow_test.go:26]:
```suggestion
func TestMaintenanceWindowAgentUpgrade(t *testing.T) {
```

File: lib/services/local/generic/nonce.go [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+// NonceViolation is the error returned by FastUpdateNonceProtectedResource when a nonce-protected
+// update fails due to concurrent modification. This error should be caught and re-mapped into an
+// appropriate user-facing message for the given resource type.
+var NonceViolation = errors.New("nonce-violation")
+
+// NonceProtectedResource describes the expected methods for a resource that is protected
+// from concurrent modification by a nonce.
+type NonceProtectedResource interface {
```
Comment (smallinsky) [inline lib/services/local/generic/nonce.go:44]: Can we make this interfaces unexported. It looks like it is used only inside this package.

File: api/proto/teleport/legacy/client/proto/authservice.proto [hunk 9]
```diff
@@ -2242,6 +2242,31 @@ message UpdateHeadlessAuthenticationStateRequest {
...
+  // CanonicalSchedule is the teleport-facing schedule repr.
+  types.AgentUpgradeSchedule CanonicalSchedule = 1;
+  // KubeControllerSchedule encodes upcoming upgrade upgrade windows in a format known
+  // to the kube upgrade controller. Teleport agents should treat this value as an
+  // opaque blob.
+  string KubeControllerSchedule = 2;
+  // SystemdUnitSchedule encodes the upcoming upgrade windows in a format known to
+  // the teleport-upgrade systemd unit. Teleport agents should treat this value as an
+  // opaque blob.
+  string SystemdUnitSchedule = 3;
```
Comment (smallinsky) [inline api/proto/teleport/legacy/client/proto/authservice.proto:2267]: Why we bother about a client formatting ?  If a client is able to call this gRPC method is should be also to convert the `AgentUpgradeSchedule` and `SystemdUnitSchedule` proto object to the proper format on client side.

File: api/types/maintenancewindow.go (hunk 4 above)
Comment (fspmarshall) [reply api/types/maintenancewindow.go:63]: It won't be larger than 23 unless there is a bug in `CheckAndSetDefaults`.  This is just for the purposes of limiting the impact of such a bug by ensuring that we won't arbitrarily jump around/skip days if an invalid hour value does make it in somehow.  go's time package will just move the date arbitrarily forward to accommodate a larger than 23 hour value.  A bug in how we handle `UTCStartHour` resulting in the wrong hour being used in the timestamp is easy to debug... a bug resulting in wacky timestamps way off from the expected day is harder to debug.

Basically, I just do this cuz I dislike how go's time package handles bad inputs.

File: api/proto/teleport/legacy/types/types.proto (hunk 3 above)
Comment (fspmarshall) [reply api/proto/teleport/legacy/types/types.proto:5143]: Yes, tho making sure enums serialize consistently across json/yaml/logs in a user-friendly way has been painful in the past.  Unless wire efficiency is a big priority (it isn't here, since this is a singleton only ever consumed by auth), I prefer strings.  Much easier to get consistent user-friendly representations.

File: api/proto/teleport/legacy/client/proto/authservice.proto (hunk 9 above)
Comment (fspmarshall) [reply api/proto/teleport/legacy/client/proto/authservice.proto:2267]: One of the core design goals of this system was to minimize, to the greatest extent possible, the potential for agent-side bugs to prevent upgrades.  If the schedule format is bugged, we want to be able to fix it by updating the control plane _only_.

What we want to avoid is a scenario where some bug forces a user to manually upgrade because of a problem in the teleport agent code.  So long as the schedule is an opaque blob, we can be reasonably certain that the agent will either successfully export the schedule (in which case any bugs can be fixed by just updating auth), or will fail to update the schedule (which the upgraders interpret as a sign that the agent is unhealthy and should be upgraded).

If, on the other hand, the agent is the one encoding the schedule and encodes it in a bugged way (e.g. causing the updater to crash when it reads the schedule), then we might get "stuck", and need manual intervention on a per-agent basis.

File: api/proto/teleport/legacy/types/types.proto [hunk 10]
```diff
@@ -5132,6 +5132,67 @@ message KubernetesResourceSpecV1 {
...
+}
+
+// ScheduledAgentUpgradeWindow is a dervied value representing a single
+// upgrade window. Generated by the AgentUpgradeWindow config object.
+message ScheduledAgentUpgradeWindow {
```
Comment (zmb3) [inline api/proto/teleport/legacy/types/types.proto:5170]: I don't quite understand the comment here. What's the difference between `AgentUpgradeWindow` and `ScheduledAgentUpgradeWindow`?

File: api/types/constants.go [hunk 11]
```diff
@@ -338,6 +338,13 @@ const (
 	// KindHeadlessAuthentication is a headless authentication resource.
 	KindHeadlessAuthentication = "headless_authentication"
 
+	// KindMaintenanceWindow determines maintenance times for the cluster.
+	KindMaintenanceWindow = "maintenance_window"
```
Comment (zmb3) [inline api/types/constants.go:342]: Are we using _maintenance window_ and _upgrade window_ interchangeably or are they separate things? If they are the same thing, I think it would be best to use the same name consistently.

File: api/types/maintenancewindow.go [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+	}
+
+	return nil
+}
+
+func (m *MaintenanceWindowV1) GetNonce() uint64 {
+	return m.Nonce
+}
+
+func (m *MaintenanceWindowV1) WithNonce(nonce uint64) any {
```
Comment (zmb3) [inline api/types/maintenancewindow.go:206]: Why do you erase the type here?

File: lib/auth/auth.go [hunk 13]
```diff
@@ -643,6 +660,71 @@ func (a *Server) SetHeadlessAuthenticationWatcher(headlessAuthenticationWatcher
...
+	}
+
+	rawStartHour, err := getter(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	startHour := uint32(rawStartHour)
+	if int64(startHour) != rawStartHour {
+		return trace.Errorf("faild to convert start hour %d to uint32 (this is a bug)", rawStartHour)
```
Comment (zmb3) [inline lib/auth/auth.go:678]: Why not just check for a start hour in [0, 24)?

This seems like a complicated check for something with relatively few possible valid values.

File: lib/auth/auth_with_roles.go [hunk 14]
```diff
@@ -5828,6 +5828,38 @@ func (a *ServerWithRoles) CloneHTTPClient(params ...roundtrip.ClientParam) (*HTT
...
+
+func (a *ServerWithRoles) GetMaintenanceWindow(ctx context.Context) (types.MaintenanceWindow, error) {
+	if err := a.action(apidefaults.Namespace, types.KindMaintenanceWindow, types.VerbRead); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return a.authServer.GetMaintenanceWindow(ctx)
+}
+
+func (a *ServerWithRoles) UpdateMaintenanceWindow(ctx context.Context, mw types.MaintenanceWindow) error {
```
Comment (zmb3) [inline lib/auth/auth_with_roles.go:5849]: Will this be surfaced in the UI?

File: lib/backend/kubernetes/kubernetes.go [hunk 15]
```diff
@@ -61,12 +65,16 @@ type Config struct {
 	// Namespace is the Agent's namespace
 	// Field is required
 	Namespace string
-	// SecretName is unique secret per agent where state and identity will be stored.
+	// SecretName uniquely identifies the secret. Conventionally this will be set
```
Comment (zmb3) [inline lib/backend/kubernetes/kubernetes.go:68]: What is the secret though? The godoc refers to the secret as if the reader should already know what the secret is for.

**Review by zmb3 — APPROVED** [review]

**Review by smallinsky — APPROVED** [review]
I have left some nit comments but beside that LGTM.

File: api/types/constants.go (hunk 11 above)
Comment (fspmarshall) [reply api/types/constants.go:345]: Naming is hard. The resource is called "maintenance window".  The only currently supported block is `agent_upgrades`.  Ex:

```yaml
kind: maintenance_window
spec:
  agent_upgrades:
    utc_start_hour: 3
    weekdays:
    - Mon
    - Wed
    - Fri
```

The idea, however, is to allow _other_ maintenance window sub-fields to be added in the future, so there is an "agent upgrade maintenance window" that exists now, but we might add other maintenance window options in the future which might not cover the same time range or purpose, but would still be part of the "maintenance window" config object.  I was originally going to name the top-level resource `ClusterSchedule`, but that felt _too_ generic.

File: api/types/maintenancewindow.go (hunk 12 above)
Comment (fspmarshall) [reply api/types/maintenancewindow.go:214]: I did this because I wanted to include `WithNonce` in the `MaintenanceWindow` interface.  We could alternatively implement the `NonceProtectedResource` interface with a generic parameter.  Ex:

```golang
type NonceProtectedResource[T any] interface {
	GetNonce() uint64
	WithNonce(uint64) T
	// ...other methods
}


// example usage
func Increment[T NonceProtectedResource[T]](rsc T) T {
	return rsc.WithNonce(rsc.GetNonce() + 1)
}

```

This would eliminate the need to do type-casting inside functions that are generic over the `NonceProtectedResource` interface, but would hide the fact that `MaintenanceWindow` types are expected to implement `WithNonce`.

IDK what the right answer is here really. Typically when I see `any` in a signature, I take it as a warning that I need to be careful about my concrete types, because some type-casting is going to happen somewhere.  Because of that, I opted to take the `WithNonce(uint64) any` route... just because it at least forces an implementer of the `MaintenanceWindow` interface to _confront_ the fact that their type is going to be expected to play nicely in a context where type-casting is happening.

File: lib/auth/auth.go (hunk 13 above)
Comment (fspmarshall) [reply lib/auth/auth.go:678]: Valid time range is checked later in `CheckAndSetDefaults`.  I just prefer to have a separate error specially for weird conversion issues... in this specific case it likely doesn't make a difference, but IMO it's good practice to call out failed conversions explicitly, since errors produced by integer conversion issues can be weird.

File: lib/auth/auth_with_roles.go (hunk 14 above)
Comment (fspmarshall) [reply lib/auth/auth_with_roles.go:5853]: Cloud users already have a UI for editing the "cloud maintenance start hour", which this resource is automatically derived from.

For on-prem users this will be a cli-only resource for now.

File: api/types/maintenancewindow.go (hunk 12 above)
Comment (zmb3) [reply api/types/maintenancewindow.go:214]: I feel like we are overcomplicating things here. Do we really need generic interfaces to build an API for setting a maintenance window? If we were writing this code 18 months ago and generics were not available what would we have done instead?

File: api/types/constants.go (hunk 11 above)
Comment (zmb3) [reply api/types/constants.go:345]: If the resource is called _maintenance window_, why cant we use that name in the protos instead of "upgrade window"?

File: lib/auth/auth.go (hunk 13 above)
Comment (zmb3) [reply lib/auth/auth.go:678]: We use integers all over the place in our protos and I'm not aware of any other checks like this. It took me a few minutes of reading this to understand what was going on.

My vote would be to simply remove it and just check `if startHour >= 0 && startHour < 24`.

Comment (fspmarshall) [reply lib/auth/auth.go]: Alright, removed this check so it'll just be caught by the normal `CheckAndSetDefaults` call.

**Comment by zmb3** [comment]
Removing approval as flaky test detector shows that `TestNonceParallelism` deadlocks.

File: api/types/maintenancewindow.go (hunk 12 above)
Comment (fspmarshall) [reply api/types/maintenancewindow.go:214]: > Do we really need generic interfaces to build an API for setting a maintenance window?

No, but we've started building up some useful tooling in `local/generic` to make working with resource CRUD a bit simpler.  I don't see why we wouldn't want to make useful patterns available there for future reuse if we have the option.

> If we were writing this code 18 months ago and generics were not available what would we have done instead?

Not much different... replace `T` with `*types.MaintenanceWindowV1` and get rid of a three-method interface.  The shallow copying method would _still_ be doing type-erasure... it would return its parent interface (`types.MaintenanceWindow`) instead of `any`, just like the `Clone()` methods currently do, but that would still mean us returning an interface type, and would still mean we'd need to downcast to a concrete type if we wanted to be sure the type we are writing into the backend is the same as the type we are expecting to read.
